internal/cfgmgr/cfgmgr.go

```diff
@@ -0,0 +1,149 @@
+
+	STRING_BUFFER_SIZE = uint64(1024)
+)
+
+func getDevPKeyDeviceLocationPaths() (*devPropKey, error) {
+	guid, err := guid.FromString(DEVPKEY_LOCATIONPATHS_GUID)
+	if err != nil {
+		return nil, err
+	}
+	return &devPropKey{
+		fmtid: guid,
+		pid:   37,
+	}, nil
+}
+
+func GetDeviceLocationPathsFromIDs(ids []string) ([]string, error) {
+	result := []string{}
+	devPKeyDeviceLocationPaths, err := getDevPKeyDeviceLocationPaths()
+	if err != nil {
+		return nil, err
+	}
+	for _, id := range ids {
+		var devNodeInst uint32
+		err = cmLocateDevNode(&devNodeInst, id, CM_LOCATE_DEVNODE_NORMAL)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to locate device node for %s", id)
+		}
+		propertyType := uint64(0)
+		propertyBufferSize := STRING_BUFFER_SIZE
+		var propertyBuffer [STRING_BUFFER_SIZE]uint16
+		err = cmGetDevNodeProperty(devNodeInst, devPKeyDeviceLocationPaths, &propertyType, &propertyBuffer[0], &propertyBufferSize, 0)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get location path property from device node for %s with", id)
+		}
+		if propertyType != DEVPROP_TYPE_STRING_LIST {
+			return nil, fmt.Errorf("expected to return property type DEVPROP_TYPE_STRING_LIST %d, instead got %d", DEVPROP_TYPE_STRING_LIST, propertyType)
+		}
+		if propertyBufferSize > STRING_BUFFER_SIZE {
+			return nil, fmt.Errorf("location path is too long, size %d", propertyBufferSize)
+		}
+		formattedResult := convertFirstNullTerminatedValueToString(propertyBuffer[:propertyBufferSize])
+		result = append(result, formattedResult)
+	}
+
+	return result, nil
+}
+
+// helper function that finds the first \u0000 rune and returns the wide string as a regular go string
+func convertFirstNullTerminatedValueToString(buf []uint16) string {
+	r := utf16.Decode(buf)
+	converted := string(r)
+	zerosIndex := strings.IndexRune(converted, '\u0000')
+	if zerosIndex != -1 {
+		converted = converted[:zerosIndex]
+	}
+	return converted
+}
+
+func GetChildrenFromInstanceIDs(parentIDs []string) ([]string, error) {
+	var result []string
+	for _, id := range parentIDs {
+		var devNodeInst uint32
+		if err := cmLocateDevNode(&devNodeInst, id, CM_LOCATE_DEVNODE_NORMAL); err != nil {
+			return nil, err
+		}
+		pszFilterParentID := []byte(id)
+		children, err := getDeviceIDList(&pszFilterParentID[0], CM_GETIDLIST_FILTER_BUSRELATIONS)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, children...)
+	}
+	return result, nil
+}
+
+func getDeviceIDList(pszFilter *byte, ulFlags uint64) ([]string, error) {
+	listLength := uint32(0)
+	if err := cmGetDeviceIDListSize(&listLength, pszFilter, ulFlags); err != nil {
+		return nil, err
+	}
+	if listLength == 0 {
+		return []string{}, nil
+	}
+	buf := make([]byte, uint64(listLength))
+	if err := cmGetDeviceIDList(pszFilter, &buf[0], uint64(listLength), ulFlags); err != nil {
+		return nil, err
+	}
+
+	var result []string
+	prev := 0
+	for i, c := range buf {
```

Comment:
From the docs , it sounds like the buffer size could be larger than the actual data container within, so we can't rely on enumerating the entire buffer looking for null terminated strings. Rather, CM_Get_Device_ID_List will set an extra NULL at the end of the list, and we should look for that to determine when to stop enumerating.

https://docs.microsoft.com/en-us/windows/win32/api/cfgmgr32/nf-cfgmgr32-cm_get_device_id_lista
https://docs.microsoft.com/en-us/windows/win32/api/cfgmgr32/nf-cfgmgr32-cm_get_device_id_list_sizea

Reply:
Updated. Please take a look.